src/gun_http2.erl

```diff
@@ -438,7 +447,12 @@ data_frame1(State0, StreamID, IsFin, Data, EvHandler, EvHandlerState0,
 					{update_window(State1), EvHandlerState1}
 			end
 	end,
-	{maybe_delete_stream(State, StreamID, remote, IsFin), EvHandlerState}.
+	case StateOrError of
+		{state, State} ->
+			{maybe_delete_stream(State, StreamID, remote, IsFin), EvHandlerState};
+		Error={error, _} ->
+			{Error, EvHandlerState}
+	end.
```

Comment:
We should probably maybe_delete_stream regardless of the error just so we keep the state consistent.

Reply:
Of course we don't yet have errors here but we will in update_window which may happen even after the stream has terminated and so we should update the state.

Reply:
To be clear, no changes needed here (`data_frame1`) since it's the same situation as in `send_data` (the other comment), right?

Reply:
No this one is needed because it's NOT the same situation: the stream has really ended, all data for the stream is sent/received, but a window_update failed after. In send_data it is the sending of the stream's data that failed so the stream didn't end and if there's a problem we probably want to know about that stream.

Reply:
OK, thanks for explaining.

If we return `{state, maybe_delete_stream(State0, ...)}` even in the error case, what to do with the error? Ignore it?

Reply:
Would it be a problem to return `[{state, maybe...(State)}, {error, Reason}]`? If it's a problem, please just add a `@todo` comment for me to fix later.

Reply:
It seems to work. Thx.

Reply:
Hrm, it doesn't work, because of code matching on state or error, like below. I'm adding a todo instead.

```Erlang
parse(Data, State0=#http2_state{status=Status, http2_machine=HTTP2Machine, streams=Streams},
		CookieStore0, EvHandler, EvHandlerState0) ->
	MaxFrameSize = cow_http2_machine:get_local_setting(max_frame_size, HTTP2Machine),
	case cow_http2:parse(Data, MaxFrameSize) of
		{ok, Frame, Rest} ->
			case frame(State0, Frame, CookieStore0, EvHandler, EvHandlerState0) of
				{Error={error, _}, CookieStore, EvHandlerState} ->
					{Error, CookieStore, EvHandlerState};
				{{state, State}, CookieStore, EvHandlerState} ->
					parse(Rest, State, CookieStore, EvHandler, EvHandlerState)
			end;
```

Reply:
I can normalize the result of frame/1 using a function like `commands/N` found in other modules. Is this what you have in mind?

```Erlang

			{Commands, CookieStore, EvHandlerState} = frame(State0, Frame,
				CookieStore0, EvHandler, EvHandlerState0),
			case commands(Commands, State0) of
				{state, State} ->
					parse(Rest, State, CookieStore, EvHandler, EvHandlerState)
				ResCommands ->
					{ResCommands, CookieStore, EvHandlerState};
			end;
```

Reply:
It might end up necessary yes. If you do that please keep it as a separate commit.

Reply:
I can save it for another PR, if you think this one can be merged as is.

Reply:
I've pushed this change as a commit in separate branch (on top of this PR): https://github.com/Nordix/gun/commit/9340a1cbc28eed7c062daa72f7a0f9e49414ddaa

If you like it, I can add id here in this PR. (I'm not sure if I like it or not.)

Reply:
Hm that's not what I had in mind. Let's just keep the todo for now. Review was good this time so I'll proceed with merge now.